### PR TITLE
Win package

### PR DESCRIPTION
**Pull request by Pat-Hof**
Adds 2 options to the Win_package module:

- display_name
- display_version

It permit either the display_name of a software or the display_name and the display_version to check its installation state.

How its works : 

If the name and optionally the version of a software are present in the x86 or x64 Uninstall registry trees, its state is installed. if found, it gets the uninstall string for deinstallation.

The display_name and optionally the display_version can be found in the Programs and Features Panel 

Examples :
```yaml
- name: Install 7zip and use the display_name for the installation check found in
Programs and Features Panel
  win_package:
    path: C:\temp\7zip.exe
    display_name: 7-Zip 16.04 (x64 edition)

- name: Install Zotero use the displayname and displayversion
   win_package:
     path:  C:\temp\Zotero-5.0.52_setup.exe
     arguments: /S
     display_name: Zotero
     display_version: 5.0.52
```

##### ISSUE TYPE
 
- Feature Pull Request

##### COMPONENT NAME

win_package

##### ANSIBLE VERSION
ansible 2.5.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15 (default, May  1 2018, 05:55:50) [GCC 7.3.0]

##### ADDITIONAL INFORMATION
Tested on Windows 10 platform and for the error conditions as well.

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/73315/22/tests

**Comment by dagwieers** [comment]
Beware that [the deadline](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_7.rst#release-schedule) for getting new features/modules accepted in Ansible v2.7 is nearing, it is set to either 2018-08-23 or 2018-08-30. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to the [Windows Working Group meeting agenda](https://github.com/ansible/community/issues/294) to get it resolved.

File: lib/ansible/modules/windows/win_package.ps1 [hunk 1]
```diff
@@ -221,12 +227,21 @@ Function Get-ProgramMetadata($state, $path, $product_id, $credential, $creates_p
...
-        } elseif ($creates_path -eq $null -and $creates_service -eq $null) {
-            # we need to fail without the product id at this point
-            Fail-Json $result "product_id is required when the path is not an MSI or the path is an MSI but not local"
+        } elseif ($creates_path -eq $null -and $creates_service -eq $null -and $display_name -eq $null) {
+            # we need to fail without the product id or display_name at this point
+            Fail-Json $result "product_id or display_name is required when the path is not an MSI or the path is an MSI but not local"
         }
     }
 
+    if ($display_name -ne $null) {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_package.ps1]: You may as well set this when `$metadata` is created, the value will still be `$null` when `$display_name/version` is also null

File: lib/ansible/modules/windows/win_package.ps1 [hunk 2]
```diff
@@ -221,12 +227,21 @@ Function Get-ProgramMetadata($state, $path, $product_id, $credential, $creates_p
...
+        $metadata.display_name = $display_name
+    }
+
+    if ($display_version -ne $null) {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_package.ps1]: Same as above, no need for an if statement just set display_version in `$metadata` to `$display_version`

File: lib/ansible/modules/windows/win_package.ps1 [hunk 3]
```diff
@@ -248,11 +263,65 @@ Function Get-ProgramMetadata($state, $path, $product_id, $credential, $creates_p
...
+                $uninstall_key  = $uninstall_path
+                $metadata.product_id = $product_id
+                if ( $metadata.display_version -ne $null -and $metadata.display_version -eq $displayversion) {
+                    $metadata.installed = $true
+                } else {
+                    $metadata.installed = $true
+                }
+            }
+        }
+        foreach ($product_id in $uninstall_array_wow64) {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_package.ps1]: Just join the 2 arrays together and run the loop on all the entries.

**Review by jborean93 — CHANGES_REQUESTED** [review]
It would be great if you could add some integration tests, feel free to reach out if you are unsure on how to do this.

File: lib/ansible/modules/windows/win_package.ps1 (hunk 1 above)
Comment (Pat-Hof) [reply lib/ansible/modules/windows/win_package.ps1]: Hi!
Sorry for the long silence
Changes done

File: lib/ansible/modules/windows/win_package.ps1 (hunk 2 above)
Comment (Pat-Hof) [reply lib/ansible/modules/windows/win_package.ps1]: Changes done

File: lib/ansible/modules/windows/win_package.ps1 (hunk 3 above)
Comment (Pat-Hof) [reply lib/ansible/modules/windows/win_package.ps1]: done

**Comment by Pat-Hof** [comment]
Hi,

> It would be great if you could add some integration tests, feel free to reach out if you are unsure on how to do this.

Where can I find  some documentation/hints.example for writing this tests ? I've never done something before.

Best Regards

**Comment by mattclay** [comment]
@Pat-Hof Do you need help fixing your pull request? It currently contains many unrelated changes.

**Comment by mattclay** [comment]
Closing to avoid polluting the [by file report](https://ansible.sivel.net/pr/byfile.html).

Feel free to open a new PR, comment here to ask for help, or ask on #ansible-devel on Freenode about rebasing a PR. The [rebasing documentation](https://docs.ansible.com/ansible/devel/dev_guide/developing_rebasing.html) may also be helpful.

**Pull request closed without merging**